services:
  postgres:
    build:
      context: ./postgres
      dockerfile: ./Dockerfile
    image: djangoblog_e2e_postgres_img
    container_name: djangoblog_e2e_postgres
    restart: always
    env_file:
      - ./.envs/e2e/postgres
    volumes:
      - db_data_e2e:/var/lib/postgresql/data
    ports:
      - "5432:5432"
    healthcheck:
      test: ["CMD", "/usr/local/bin/check-db.sh"]
      interval: 5s
      timeout: 5s
      retries: 5

  backend:
    build:
      context: .
      dockerfile: ./backend/dockerfile/local/Dockerfile
    working_dir: /app
    image: djangoblog_e2e_backend_img
    container_name: djangoblog_e2e_backend
    restart: always
    user: "${UID:-1000}:${GID:-1000}"
    depends_on:
      postgres:
        condition: service_healthy
    env_file:
      - ./.envs/e2e/django
      - ./.envs/e2e/postgres
    environment:
      - DJANGO_SETTINGS_MODULE=config.settings_fortest  # ← 本番ではないテスト専用設定
    volumes:
      - ./backend:/app
    ports:
      - "8000:8000"
    command: python manage.py runserver 0.0.0.0:8000

  frontend:
    build:
      context: ./frontend
      dockerfile: ./dockerfile/local/Dockerfile
    image: djangoblog_e2e_frontend_img
    container_name: djangoblog_e2e_frontend
    depends_on:
      - backend
    env_file:
      - ./.envs/e2e/frontend
    volumes:
      - ./frontend:/app
      - node_modules:/app/node_modules
    ports:
      - "5173:5173"
    command: npm run dev --host
    user: "${UID:-1000}:${GID:-1000}"
    extra_hosts:
      - "localhost:host-gateway"  # ← fetch("http://localhost:8000/...") 対応
  # e2e専用環境だが、
  playwright-real:
    build:
      context: .
      dockerfile: ./playwright-real/dockerfile/local/Dockerfile
    image: djangoblog_e2e_playwright_real_img
    container_name: djangoblog_e2e_playwright_real
    depends_on:
      - frontend
      - backend
    working_dir: /app
    volumes:
      - ./playwright-real/e2e_tests:/app/e2e_tests
      - ./backend:/app/backend
    env_file:
      - ./.envs/e2e/django
      - ./.envs/e2e/postgres  # ← これが抜けていたのが原因
    environment:
      - DJANGO_SETTINGS_MODULE=config.settings_fortest
      - DJANGO_ALLOW_ASYNC_UNSAFE=1
      - PYTHONPATH=/app/backend  # ← backend 配下のモジュールをインポート可能に
    extra_hosts:
      - "localhost:host-gateway"
    user: "${UID:-1000}:${GID:-1000}"
    # entrypoint: [ "tail", "-f", "/dev/null" ]  # ← デバッグ時に便利

volumes:
  db_data_e2e:
  node_modules:
